### PR TITLE
Revert the revert of Shared Directory changes and make them non breaking in main

### PR DESCRIPTION
**Pull request by jatgarg**
1.) These changes were already merged in main but needed to be reverted as per the decision taken then because it leads to OWH change their implementation of IDirectory.
2.) As per the latest discussion, we can introduce these changes again in non breaking way and mention in upcoming changes.

Previous merged PRs: 
https://github.com/microsoft/FluidFramework/pull/9087 - Add dispose event on directory deletion
https://github.com/microsoft/FluidFramework/pull/9451 - Emit dispose event recursively

### CODE REVIEW

File: packages/dds/map/package.json [hunk 1]
```diff
@@ -61,6 +61,7 @@
   "dependencies": {
     "@fluidframework/common-definitions": "^0.20.1",
     "@fluidframework/common-utils": "^0.32.1",
+    "@fluidframework/container-utils": "^0.58.1000",
```
Comment (heliocliu) [inline packages/dds/map/package.json:64]:
```suggestion
    "@fluidframework/container-utils": "^0.58.3000",
```

File: api-report/map.api.md [hunk 2]
```diff
@@ -37,7 +38,7 @@ export class DirectoryFactory {
 }
 
 // @public
-export interface IDirectory extends Map<string, any>, IEventProvider<IDirectoryEvents> {
+export interface IDirectory extends Map<string, any>, IEventProvider<IDirectoryEvents>, IDisposable {
```
Comment (anthony-murphy) [inline api-report/map.api.md:41]: this is a breaking change. it needs to be Partial<IDisposable> to be non-breaking

File: api-report/map.api.md [hunk 3]
```diff
@@ -171,6 +174,10 @@ export class SharedDirectory extends SharedObject<ISharedDirectoryEvents> implem
     createSubDirectory(subdirName: string): IDirectory;
     delete(key: string): boolean;
     deleteSubDirectory(subdirName: string): boolean;
+    // (undocumented)
+    dispose(error?: Error): void;
```
Comment (anthony-murphy) [inline api-report/map.api.md:178]: i would expect these to be optional. might need to make them explicitly optional in their signatures

Comment (jatgarg) [reply api-report/map.api.md:178]: So, implementing Partial does implicitly imply that these could be optional. So the class implementing this interface, may/may not implement these apis and in both cases it should be fine.

Comment (heliocliu) [reply api-report/map.api.md:178]: could just be an issue with the docs generation?

**Review by heliocliu — APPROVED** [review]

**Pull request merged**